sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs

```diff
@@ -659,604 +600,482 @@ protected EventProcessorClient()
+                    {
+                       FullyQualifiedNamespace = FullyQualifiedNamespace,
+                       EventHubName = EventHubName,
+                       ConsumerGroup = ConsumerGroup,
+                       PartitionId = partition,
+                       StartingPosition = PartitionStartingPositionDefaults[partition]
+                    };
                 }
             }
-
-            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
-
-            var closingEventArgs = new PartitionClosingEventArgs(partitionId, reason, cancellationToken);
-            _ = OnPartitionClosingAsync(closingEventArgs);
         }
 
         /// <summary>
-        ///   Signals the <see cref="EventProcessorClient" /> to begin processing events. Should this method be called while the processor is running, no action is taken.
+        ///   Invokes a specified action only if this <see cref="EventProcessorClient" /> instance is not running.
         /// </summary>
         ///
-        /// <param name="async">When <c>true</c>, the method will be executed asynchronously; otherwise, it will execute synchronously.</param>
-        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the start operation.  This won't affect the <see cref="EventProcessorClient" /> once it starts running.</param>
+        /// <param name="action">The action to invoke.</param>
         ///
-        /// <exception cref="EventHubsException">Occurs when this <see cref="EventProcessorClient" /> instance is already closed.</exception>
-        /// <exception cref="InvalidOperationException">Occurs when this method is invoked without <see cref="ProcessEventAsync" /> or <see cref="ProcessErrorAsync" /> set.</exception>
+        /// <exception cref="InvalidOperationException">Occurs when this method is invoked while the event processor is running.</exception>
         ///
-        private async Task StartProcessingInternalAsync(bool async,
-                                                        CancellationToken cancellationToken)
+        private void EnsureNotRunningAndInvoke(Action action)
         {
-            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            var releaseGuard = false;
 
-            if (ActiveLoadBalancingTask == null)
+            if (!IsRunning)
             {
-                if (async)
-                {
-                    await RunningTaskSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
-                }
-                else
-                {
-                    RunningTaskSemaphore.Wait(cancellationToken);
-                }
-
                 try
                 {
-                    cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
-
-                    lock (StartProcessorGuard)
+                    if (ProcessorStatusGuard.Wait(TimeSpan.FromSeconds(1)))
```

Comment:
It's definitely something worth talking over.   The rationale was to prevent things from hanging indefinitely when attempting to add/remove a handler.   Since those calls are synchronous and we don't have control over the signature, using a cancellation token wasn't possible.   Having the handler registration calls take more than a second seemed undesirable in the event that there was a lock convoy or something forming.

That said, it is entirely possible that I'm overthinking things here and we should just wait for things to resolve on their own.   I'd appreciate any thoughts that you have.

Reply:
> That said, it is entirely possible that I'm overthinking things here and we should just wait for things to resolve on their own. I'd appreciate any thoughts that you have.

I think removing the timeouts here is worthwhile.  The case this seems to be protecting against is the case where `StartProcessingAsync` takes a while and holds the lock while doing so, while at the same time some other thread is trying to register a handler.  Having a long wait here doesn't feel like the worst thing in the world.  In practice, I think we don't expect setting up the event handlers and calling `StartProcessingAsync` to happen concurrently, and so we shouldn't worry about long blocking here.

In short, I think if you ever observed this weird case, your code was unlikely to be very robust, since it inherently has setting up your event handlers racing with calling `StartProcessingAsync` which feels very much like an anti-pattern.

I'd personally bias towards the simpler code, unless there's some reasonable case where these operations could be interleaved.

Reply:
Fair enough;  your rationale makes sense.   Seems that I'm guilty of overthinking this one.  I'll pull them out.